src/OrbitQt/CallTreeWidget.cpp

```diff
@@ -170,28 +170,87 @@ void CallTreeWidget::ResizeColumnsIfNecessary() {
   column_resizing_state_ = ColumnResizingState::kDone;
 }
 
-static std::string BuildStringFromIndices(const QModelIndexList& indices) {
-  std::string buffer;
+static std::string BuildStringFromIndices(QTreeView* tree_view, const QModelIndexList& indices) {
+  // The order of rows in the selection model is not the order in which the rows are displayed.
+  // We can recover that order by sorting using as comparison function for two rows the
+  // lexicographic order of the vector of the row indices from the root to those rows.
+  struct ItemWithAncestorRows {
+    enum ExpandedState { kLeaf = 0, kExpanded, kCollapsed };
+
+    std::string data;
+    bool is_first_in_row = false;
+    ExpandedState expanded_state = kLeaf;
+    std::list<int> ancestor_rows;
+
+    bool operator<(const ItemWithAncestorRows& that) const {
+      return ancestor_rows < that.ancestor_rows;
+    }
+  };
+  std::vector<ItemWithAncestorRows> items;
+
   std::optional<QModelIndex> prev_index;
-  // Note: indices are sorted by row in order of selection and then by column in ascending order.
+  // Note: the indices vector is sorted by row in order of selection and then by column in ascending
+  // order.
   for (const QModelIndex& index : indices) {
-    if (prev_index.has_value()) {
-      // row() is the position among siblings: also compare parent().
-      if (index.row() != prev_index->row() || index.parent() != prev_index->parent()) {
-        buffer += "\n";
-      } else {
-        buffer += ", ";
-      }
+    ItemWithAncestorRows item;
+    item.data = index.data().toString().toStdString();
+
+    // row() is the position among siblings: also compare parent().
+    item.is_first_in_row = !prev_index.has_value() || index.row() != prev_index->row() ||
+                           index.parent() != prev_index->parent();
+
+    if (index.model()->rowCount(index) == 0) {
+      item.expanded_state = ItemWithAncestorRows::kLeaf;
+    } else if (tree_view->isExpanded(index)) {
+      item.expanded_state = ItemWithAncestorRows::kExpanded;
+    } else {
+      item.expanded_state = ItemWithAncestorRows::kCollapsed;
     }
-    buffer += index.data().toString().toStdString();
+
+    for (QModelIndex temp_index = index; temp_index.isValid(); temp_index = temp_index.parent()) {
+      item.ancestor_rows.emplace_front(temp_index.row());
+    }
+
+    items.emplace_back(std::move(item));
     prev_index = index;
   }
+
+  // Use stable_sort to maintain the order in the same row (i.e., across columns).
+  std::stable_sort(items.begin(), items.end());
+  std::string buffer;
+  for (size_t i = 0; i < items.size(); ++i) {
+    const ItemWithAncestorRows& item = items[i];
+    if (item.is_first_in_row) {
+      if (i > 0) {
+        buffer.append("\n");
+      }
+
+      for (size_t indent = 0; indent < item.ancestor_rows.size() - 1; ++indent) {
+        buffer.append("  ");
+      }
```

Comment:
What about `buffer.append(2 * item.ancestor_rows.size(); ' ');`?

Reply:
Much better. By the way, I'm using `item.ancestor_rows.size() - 1`, excluding the actual row of the element from the count.

Reply:
Yes, I missed that! Thanks!